### PR TITLE
iDeep test

### PR DESCRIPTION
**Pull request by niboshi**
This is a PR for testing iDeep: https://github.com/chainer/chainer/pull/3865

~Currently it tries to install iDeep via GitHub, but it will fail.
I will revise the PR when iDeep will be released on PyPI.~  <- Done

### CODE REVIEW

File: docker.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ def get_numpy_choices():
 
 RUN yum -y update && \\
     yum -y install epel-release && \\
-    yum -y install gcc gcc-c++ git kmod hdf5-devel perl make autoconf xz && \\
+    yum -y install gcc gcc-c++ cmake git kmod hdf5-devel perl make autoconf xz && \\
```
Comment (kmaehashi) [inline docker.py]: iDeep is provided as wheel, so `cmake` is not required.

**Comment by niboshi** [comment]
Looking at Jenkins test logs, it seems `pip install ideep4py` has not been not invoked. Am I doing something wrong?

File: run_test.py [hunk 2]
```diff
@@ -57,6 +57,7 @@
             'cudnn': 'cudnn51-cuda8',
             'nccl': 'nccl1.3.4',
             'protobuf-cpp': 'protobuf-cpp-3',
+            'ideep': 'ideep4py==1.0.1',
```
Comment (kmaehashi) [inline run_test.py]: `ubuntu14_py3` uses system Python (3.4), which is not supported by iDeep.

File: shuffle.py [hunk 3]
```diff
@@ -91,6 +91,10 @@ def make_conf(params):
     else:
         append_require(params, conf, 'protobuf')
 
+    if params.get('ideep') is not None:
```
Comment (kmaehashi) [inline shuffle.py]: We should exclude iDeep4Py combination if not the Python is not (2.7.5+, 3.5.2+, or 3.6.0+).
(Maybe this depends on https://github.com/chainer/chainer-test/pull/383)

File: test_cupy.sh [hunk 4]
```diff
@@ -15,7 +15,7 @@ pytest_opts=(
 if [ $CUDNN = none ]; then
   pytest_opts+=(-m 'not cudnn and not slow')
 else
-  pytest_opts+=(-m 'not slow')
+  pytest_opts+=(-m 'not ideep and not slow')
```
Comment (kmaehashi) [inline test_cupy.sh]: This script is for CuPy unittest.

File: test_cupy_slow.sh [hunk 5]
```diff
@@ -15,7 +15,7 @@ pytest_opts=(
 if [ $CUDNN = none ]; then
   pytest_opts+=(-m 'slow and not cudnn')
 else
-  pytest_opts+=(-m 'slow')
+  pytest_opts+=(-m 'slow and not ideep')
```
Comment (kmaehashi) [inline test_cupy_slow.sh]: ditto

File: run_test.py [hunk 6]
```diff
@@ -70,6 +71,7 @@
             'cuda': 'cuda90',
             'cudnn': 'cudnn7-cuda9',
             'nccl': 'nccl2.0-cuda9',
+            'ideep': 'ideep4py==1.0.1',
```
Comment (kmaehashi) [inline run_test.py]: > Looking at Jenkins test logs, it seems pip install ideep4py has not been invoked. Am I doing something wrong?

If no custom Dockerfile modification is required, add iDeep to `requires` section.
`conf` is transformed into Dockerfile like this: https://github.com/chainer/chainer-test/blob/master/docker.py#L469

File: test_cupy.sh (hunk 4 above)
Comment (niboshi) [reply test_cupy.sh]: Yes, I want to skip ideep tests in CuPy unittest.
Oh, I should have put `not ideep` in both clauses.

Comment (kmaehashi) [reply test_cupy.sh]: Oh, will iDeep test code be added to CuPy, too? (not only Chainer?)

Comment (niboshi) [reply test_cupy.sh]: Ok, I understand...

File: docker.py (hunk 1 above)
Comment (niboshi) [reply docker.py]: Fixed.

File: run_test.py (hunk 2 above)
Comment (niboshi) [reply run_test.py]: Fixed.

File: shuffle.py (hunk 3 above)
Comment (niboshi) [reply shuffle.py]: Fixed.

File: test_cupy.sh (hunk 4 above)
Comment (niboshi) [reply test_cupy.sh]: Fixed.

File: test_cupy_slow.sh (hunk 5 above)
Comment (niboshi) [reply test_cupy_slow.sh]: Fixed.

File: run_test.py (hunk 6 above)
Comment (niboshi) [reply run_test.py]: Fixed.

File: docker.py [hunk 7]
```diff
@@ -9,12 +9,18 @@
...
-    'ubuntu14_py27', 'ubuntu14_py34',
-    'ubuntu14_py35-pyenv', 'ubuntu14_py36-pyenv',
-    'ubuntu16_py27', 'ubuntu16_py35',
-    'centos6_py27-pyenv',
-    'centos7_py27', 'centos7_py34-pyenv']
+_base_choices = [
+    ('ubuntu14_py27', '2.7.6'),
+    ('ubuntu14_py34', '3.4.0'),
+    ('ubuntu14_py35-pyenv', '3.5.4'),
+    ('ubuntu14_py36-pyenv', '3.6.3'),
```
Comment (kmaehashi) [inline docker.py]: How about using `get_python_version` in https://github.com/niboshi/chainer-test/blob/585df1196a116043ea163dbbe4417f2a8556a67f/docker.py#L195-L196

Comment (niboshi) [reply docker.py]: Fixed.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**